src/rest/connection.cpp

```diff
@@ -0,0 +1,364 @@
+{
+}
+
+void Connection::Init()
+{
+    mParser = std::unique_ptr<Parser>(new Parser(&mRequest));
+    mParser->Init();
+}
+
+void Connection::Disconnect()
+{
+    mState = OTBR_REST_CONNECTION_COMPLETE;
+
+    if (mFd != -1)
+    {
+        close(mFd);
+        mFd = -1;
+    }
+}
+
+otbrError Connection::UpdateReadFdSet(fd_set &aReadFdSet, int &aMaxFd)
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    if (mState == OTBR_REST_CONNECTION_READWAIT || mState == OTBR_REST_CONNECTION_INIT)
+    {
+        FD_SET(mFd, &aReadFdSet);
+        aMaxFd = aMaxFd < mFd ? mFd : aMaxFd;
+    }
+
+    return error;
+}
+
+otbrError Connection::UpdateWriteFdSet(fd_set &aWriteFdSet, int &aMaxFd)
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    if (mState == OTBR_REST_CONNECTION_WRITEWAIT)
+    {
+        FD_SET(mFd, &aWriteFdSet);
+        aMaxFd = aMaxFd < mFd ? mFd : aMaxFd;
+    }
+
+    return error;
+}
+
+otbrError Connection::UpdateTimeout(timeval &aTimeout)
+{
+    otbrError      error = OTBR_ERROR_NONE;
+    struct timeval timeout;
+    uint32_t       timeoutLen = kReadTimeout;
+    auto           duration   = duration_cast<microseconds>(steady_clock::now() - mStartTime).count();
+
+    switch (mState)
+    {
+    case OTBR_REST_CONNECTION_READWAIT:
+        timeoutLen = kReadTimeout;
+        break;
+    case OTBR_REST_CONNECTION_CALLBACKWAIT:
+        timeoutLen = kWriteTimeout;
+        break;
+    case OTBR_REST_CONNECTION_WRITEWAIT:
+        timeoutLen = kCallbackTimeout;
+        break;
+    case OTBR_REST_CONNECTION_COMPLETE:
+        timeoutLen = 0;
+        break;
+    default:
+        break;
+    }
+
+    if (duration <= timeoutLen)
+    {
+        timeout.tv_sec  = 0;
+        timeout.tv_usec = timeoutLen - duration;
+    }
+    else
+    {
+        timeout.tv_sec  = 0;
+        timeout.tv_usec = 0;
+    }
+
+    if (timercmp(&timeout, &aTimeout, <))
+    {
+        aTimeout = timeout;
+    }
+
+    return error;
+}
+
+otbrError Connection::UpdateFdSet(otSysMainloopContext &aMainloop)
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    error = UpdateTimeout(aMainloop.mTimeout);
+    error = UpdateReadFdSet(aMainloop.mReadFdSet, aMainloop.mMaxFd);
+    error = UpdateWriteFdSet(aMainloop.mWriteFdSet, aMainloop.mMaxFd);
+
+    return error;
+}
+
+otbrError Connection::ProcessWaitRead(fd_set &aReadFdSet)
+{
+    otbrError error    = OTBR_ERROR_NONE;
+    auto      duration = duration_cast<microseconds>(steady_clock::now() - mStartTime).count();
+    int       received;
+    char      buf[2048];
+    int       err;
+
+    if (duration <= kReadTimeout)
+    {
+        if (FD_ISSET(mFd, &aReadFdSet) || mState == OTBR_REST_CONNECTION_INIT)
+        {
+            while (true)
+            {
+                received = read(mFd, buf, sizeof(buf));
+                err      = errno;
+
+                if (received <= 0)
+                {
+                    if (err != EINTR)
+                    {
+                        break;
```

Comment:
Nested `if...else...` statement is toooooo deep! Keep nesting depth low!

For example, you can write like this:
```c++
VerifyOrExit(duration <= kReadTimeout, error = OTBR_ERROR_TIMEOUT);
VerifyOrExit(FD_ISSET(mFd, &aReadFdSet) || mState == OTBR_REST_CONNECTION_INIT, OT_NOOP);

while (true) {
....
}
```

The idea is to exit as early as possible.